0005-snarkyjs-ecdsa-sha.md

```diff
@@ -0,0 +1,179 @@
+
+Sha3_224.hash(xs);
+// ..
+```
+
+### Alternative Approach
+
+Another possibility is to combine all hash functions, including Poseidon, under a shared namespace `Hash`. Developers will then be able to use these functions by calling `Hash.[hash_name].hash(xs)`. However, this would not be equivalent to the existing `Poseidon` API.
+
+However, the existing `Poseidon` API could be deprecated in favour of the new `Hash` namespace.
+
+```ts
+const Hash = {
+  Poseidon: {
+    hash: (xs: UInt8[]) => "digest",
+  },
+
+  // ..
+
+  SHA3_224: {
+    hash: (xs: UInt8[]) => "digest",
+  },
+
+  Keccak: {
+    hash: (xs: UInt8[]) => "digest",
+  },
+};
+```
+
+Developers would then be able to simply import the `Hash` namespace into their project and use all available hash functions.
+
+```ts
+import { Hash } from "snarkyjs";
+
+Hash.Poseidon.hash(xs);
+Hash.Keccak.hash(xs);
+// ..
+
+// deprecated in favor of Hash.Poseidon.hash
+Poseidon.hash(xs);
+```
+
+It is important to mention that we have a lot of freedom in the the implementation details of the API, as the gates and gadgets are already implemented in OCaml and just need to exposed and wrapped by a developer friendly API.
+
+One goal of the API is to design an easy to use interface for developers as well as maintaining consistency with already existing primitives (Poseidon in this case, Schnorr in the case of ECDSA).
+The usage of both Keccak/SHA3 and Poseidon is the same: The user passes an array of Field elements into the hash function and the output is returned. One important detail is that Keccak/SHA3 only accepts an array of Field element that each are no larger than 1 byte. We will also provide additional helper functions that allow conversion between Field element arrays and hexadecimal encoded strings.
+
+```ts
+function fieldBytesFromHex(hex: string): UInt8[];
+function hexToFieldBytes(xs: UInt8[]): string;
+```
+
+Overall, exposing new gadgets and gates follow a strict pattern that has been used all over in the SnarkyJS bindings layer. As an example, the [Poseidon](https://github.com/o1-labs/snarkyjs-bindings/blob/main/ocaml/lib/snarky_js_bindings_lib.ml#L386) implementation behaves similarly. From the point of view of SnarkyJS, these gadgets are just another set of function calls to OCaml.
+
+## ECDSA
+
+**TODO** this will be a seperate PR stacked on top of this one
+
+**Evergreen, wide-sweeping Details**
+
+**Ephemeral details**
+
+## Test plan and functional requirements
+
+## SHA3/Keccak
+
+In order to test the implementation of SHA3 and Keccak in SnarkyJS, we will follow the testing approach we already apply to other gadgets and gates.
+This includes testing the out-of- and in-snark variants using our testing framework, as well as adding a SHA3 and Keccak regression test. The regression tests will also include a set of predetermined digests to make sure that the algorithm doesn't unexpectedly change over time (similar to the tests implemented for the OCaml gadget).
```

Comment:
> a set of predetermined digests

If we're considering to utilize the test vectors from organizations like `NIST` (to ensure the correctness of the implementation), then we should specify it.

> regression test

It will be good to also mention what will form the regression testing suite (for example, functional tests, specific edge cases, errors handling, backward compatibility, etc.).

Reply:
FYI, we have a "regression test suite" in use in snarkyjs CI, which this RFC refers to. it checks that constraints generated using a given list of contracts / provable methods doesn't change: https://github.com/o1-labs/snarkyjs/blob/main/src/examples/vk_regression.ts

Reply:
+1 for specific edge cases, errors handling

Since padding is important for security and there are at least two kinds of padding supported by our Keccak gadgets, from a security perspective testing different length messages (e.g. empty, zero, one less than padding lengths, one more, multiples of, etc) are great edge-cases to test imho!

Reply:
Thanks everyone! I slightly edited this section to reflect your ideas. I am a big fan of adding (maybe not all, but at least some) of the NIST test vectors - what are your thoughts on this? Is this already being done in the crypto side? (didn't explicitly see that)